### PR TITLE
Better cache usage for previews

### PR DESCRIPTION
**Pull request by icewind1991**
Use cache data we already have instead of making filesystem calls for significant speed improvements when working with external storages

cc @DeepDiver1975 @PVince81 @georgehrke

### CODE REVIEW

**Comment by georgehrke** [comment]
also cc'ing @oparoz

**Comment by karlitschek** [comment]
makes sense :+1:

**Comment by georgehrke** [comment]
code changes look good, will test asap

File: lib/private/preview.php [hunk 1]
```diff
@@ -507,6 +508,29 @@ private function getMaxPreviewSize($allThumbnails) {
...
+	 * @param FileInfo[] $allThumbnails the list of all our cached thumbnails
+	 * @param int $width
+	 * @param int $height
+	 * @return bool
+	 */
+	private function thumbnailSizeExists($allThumbnails, $width, $height) {
+
+		foreach ($allThumbnails as $thumbnail) {
+			$name = $thumbnail['name'];
+			if (strpos($name, 'max')) {
```
Comment (oparoz) [inline lib/private/preview.php]: I don't think this is right. This only looks for the max preview.

File: lib/private/preview.php [hunk 2]
```diff
@@ -507,6 +508,29 @@ private function getMaxPreviewSize($allThumbnails) {
...
+				list($cachedWidth, $cachedHeight) = $this->getDimensionsFromFilename($name);
```
Comment (oparoz) [inline lib/private/preview.php]: This doesn't take into consideration the fact that some thumbnails are cropped and not simply resized.

**Comment by oparoz** [comment]
It's a good idea. We already have the list of thumbnails in memory, so there is no need to query the filesystem again.

**Comment by oparoz** [comment]
I see a couple of alternatives. One of them, which could be useful in quite a few situations would be to cache the thumbnail sizes in the DB. The main problem with that approach being making sure the cache gets properly cleared and updated.

File: lib/private/preview.php (hunk 1 above)
Comment (icewind1991) [inline lib/private/preview.php]: To much copy-paste

File: lib/private/preview.php (hunk 2 above)
Comment (icewind1991) [inline lib/private/preview.php]: Fixed

File: lib/private/preview.php [hunk 3]
```diff
@@ -507,6 +508,24 @@ private function getMaxPreviewSize($allThumbnails) {
 	}
 
 	/**
+	 * Check if a specific thumbnail size is cached
+	 *
+	 * @param FileInfo[] $allThumbnails the list of all our cached thumbnails
+	 * @param string $name
+	 * @return bool
+	 */
+	private function thumbnailSizeExists($allThumbnails, $name) {
```
Comment (LukasReschke) [inline lib/private/preview.php]: `array $allThumbnails`?

Comment (oparoz) [inline lib/private/preview.php]: Looks like it's correct:
https://github.com/owncloud/core/blob/7959605e4e389ee26705094656f636f23b9cfb8d/lib/private/files/view.php#L1235

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Comment by MorrisJobke** [comment]
@icewind1991 What is the stat of this? Ready to review?

**Comment by MorrisJobke** [comment]
Jenkins PR #17893

**Comment by MorrisJobke** [comment]
Jenkins is happy: https://github.com/owncloud/core/pull/17893#issuecomment-125151885

**Comment by MorrisJobke** [comment]
@georgehrke @oparoz @rullzer @Xenopathic Mind to review?

**Comment by oparoz** [comment]
- [ ] Add missing test(s)

**Comment by PVince81** [comment]
@icewind1991 can you add the missing test ?

**Comment by icewind1991** [comment]
Afaik this is covered by existing preview tests

**Comment by oparoz** [comment]
You changed the signature of setFile(), so there must be at least one new case.

**Comment by MorrisJobke** [comment]
> You changed the signature of setFile(), so there must be at least one new case.

@icewind1991 Please add the test case

**Comment by icewind1991** [comment]
Added more tests

**Comment by oparoz** [comment]
Code makes sense, tests make sense (and pass) and I didn't experience side effects.
:+1:

**Comment by MorrisJobke** [comment]
Tested and works :+1:

**Comment by MorrisJobke** [comment]
@icewind1991 please rebase

**Comment by MorrisJobke** [comment]
Tested and still works

**Pull request merged**